osutil/sync_dir.go

```diff
@@ -0,0 +1,100 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// FileState describes the expected content and meta data of a single file.
+type FileState struct {
+	Content []byte
+	Mode    os.FileMode
+}
+
+var errSameState = fmt.Errorf("file state has not changed")
+
+// EnsureDirState ensures that directory content matches expectations.
+//
+// EnsureDirState enumerates all the files in the specified directory that
+// match the provided pattern (glob). Each enumerated file is checked to ensure
+// that the contents, permissions are what is desired. Unexpected files are
+// removed. Missing files are created and differing files are corrected.  Files
+// not matching the pattern are ignored.
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
```

Comment:
"
- On a write error
 - Track in firstErr
 - Set content = nil
 - Break out of the write loop so removals are attempted on every known file.
"
I like where you are going here since it seems EnsureDirState has everything it needs to do something right, but if file 1 is ok, file 2 has a write error and file 3 would be ok, file 1 is still applied, but files 2 and 3 are deleted (assuming I understand what you are suggesting). It almost seems better to go all through the loop, tracking all the errors so that 1 and 3 are processed and 2 is deleted. This is admittedly more complicated and I'm not sure how this function is intended to be used in the grand scheme of things, so may not be appropriate.